Utilsexpand (#6788)
ethereum/web3.js: docs/docs/guides/Web3_Utils_module/Mastering_web3-utils.md

# Mastering web3.js Utils

## Introduction

## Imports

> There are three different way to import utils package.

Reviewer: ```suggestion
There are three different ways to import web3-utils package.
```

1. Import the entire web3 module and initialize a provider:

    `const [ Web3 ] = require("web3"); ` 

>     then inialiazing a provider with 

Reviewer: ```suggestion
    then initializing a provider with 
```

    `const web3 = new Web3("");`

    To be able to access the package you need to type
    `web3.utils.toWei()`

2. Import only the utils package from web3:

>     `const { utils } = require("web3");` 

Reviewer 2: it will be great to have imports TypeScript style also.  You can use tabs or just write 2 options near
```ts 
import { Web3 } from 'web3'
```

to import only utils package:
```ts 
import * as utils from 'web3-utils';
```

    Then you are also able to use any function like
    `utils.toWei()`

3. Import a specific function from the utils package: 

    `const { toWei } = require("web3-utils");`
    
    then you'll be able to use it here `toWei();`

## Random Bytes

> There are two way to generate Random Bytes.

Reviewer: ```suggestion
There are two ways to generate Random Bytes.
```

> 1. The first function that is knownn as Random Bytes, it receives a parameter of the size of bytes you want to generate and it will return the bytes in an array format.

Reviewer: ```suggestion
1. The first function that is known as Random Bytes, it receives a parameter of the size of bytes you want to generate and it will return the bytes in an array format.
```

    array format `console.log(utils.randomBytes());`

> 2. The second function is known as Randon Hex also requires a parameter of the size of bytes and it will return the btytes in Hexadecimal format

Reviewer: ```suggestion
2. The second function is known as Random Hex also requires a parameter of the size of bytes and it will return the bytes in Hexadecimal format
```

    Hex format `console.log(utils.randomHex());`

    If you don't give any arguments then both of these functions will have a default value as 32.

## Conversions - Ethereum Denominations

> We've got two different functions to perform conversions between ethereum denominations.

Reviewer: ```suggestion
We've got two different functions to perform conversions between Ethereum denominations.
```

1. `toWei(any) -> wei `

    For example, if we want to convert 1 ether to wei we can use the function below
    `console.log(utils.toWei("1", "ether"));`

    And to convert gwei to wei, we can do the following.  
    `console.log(utils.toWei("1", "gwei"));`

2. `fromWei(wei) -> any`
    To use the second function, you can convert 10^18(1000000000000000000) wei to ether, to get its output you can follow the below given methid.
    `console.log(utils.FromWei("1000000000000000000", "ether"));`

    You can follow the same method to convert 1000000000000000000 to gwei, which is given below.
    `console.log("utils.fromWei("1000000000000000000", "gwei"));`

    When you run the console, it will show us the results.

## Conversions to Hex Values

> There are many ways to convert any value into Hexadecimal but one of the most simplest and easiest is to use the function to Hex 

Reviewer: ```suggestion
There are many ways to convert any value into Hexadecimal but one of the simplest and easiest is to use the function to Hex 
```

`toHex(any) > Hex` which will accept anry parameter and will always return a Hexadecimal value so this function can accept numbers "big Int", "strings", "booleans" and "objects". Below is how you can write them.

For Numbers: 
`console.log(utils.toHex(10));`

For Big Numbers:
`console.log(utils.toHex(10n));` 

For Strings: 
`console.log(utils.toHex("10"));` 

> Strings are returned after being decoded in ASCI value, then it gets converted into a Hexadecimal.

Reviewer: ```suggestion
Strings are returned after being decoded in ASCII value, then it gets converted into a Hexadecimal.
```

`console.log(utils.numberToHex(10));`

Decimal-> Hex: 
`console.log(utils.fromDecimal(10)); ` 

String -> Hex: 
`console.log(utils.utf8ToHex("text"));` 

String -> Hex: 
`console.log(utils.asciiToHex("text"));`

bytes (array) -> Hex
Another case is if you want to convert bytes in an arrayed format
Here is how you can do it.

Array: 
`const arr = [72,12]`

Uint8Arry: 
`console.log(utils.toHex(arr));` 

Bytes: 
`console.log(utils.bytesToHex(arr));` 

## Conversions UTF and ASCII

You can convert Hexadecimal values to UTF and ASCII values. UTF has a broader range of characters and can support emojis as well, hence it is recommended to use. If you use UTF and send Emoji as a Hexadecimal value, it will return the emoji but the same case cannot be applied when it comes ASCII values.

Example function:

For Emoji to UTF:
 `console.log("utf:", utils.toUtf8("0xf09f988a"));`

For emoji to Ascii:
 `console.log("ascii:", utils.toAscii("0xf09f988a"));`

## Conversions - Numbers and Bigint

There are three different ways to convert Hexadecimal values into numbers.

1. This function `toNumber` returns Hexadecimal value in number format
Hex -> number: `console.log(utils.toNumber("0xa"));`

> 2. This function `hexToNumberString` will convert Hexadecimal value into anumber and returns it in string format

Reviewer: ```suggestion
2. This function `hexToNumberString` will convert Hexadecimal value into a number and returns it in string format
```
